internal/alb/ls/rules_comp.go

```diff
@@ -0,0 +1,185 @@
+	case conditions.FieldHTTPRequestMethod:
+		return httpRequestMethodConditionConfigMatches(desired.HttpRequestMethodConfig, current.HttpRequestMethodConfig)
+	case conditions.FieldQueryString:
+		return queryStringConditionConfigMatches(desired.QueryStringConfig, current.QueryStringConfig)
+	case conditions.FieldSourceIP:
+		return sourceIpConditionConfigMatches(desired.SourceIpConfig, current.SourceIpConfig)
+	}
+	return false
+}
+
+func hostHeaderConditionConfigMatches(desired *elbv2.HostHeaderConditionConfig, current *elbv2.HostHeaderConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+func pathPatternConditionConfigMatches(desired *elbv2.PathPatternConditionConfig, current *elbv2.PathPatternConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+func httpHeaderConditionConfigMatches(desired *elbv2.HttpHeaderConditionConfig, current *elbv2.HttpHeaderConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return (aws.StringValue(desired.HttpHeaderName) == aws.StringValue(current.HttpHeaderName)) &&
+		sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+func httpRequestMethodConditionConfigMatches(desired *elbv2.HttpRequestMethodConditionConfig, current *elbv2.HttpRequestMethodConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+func queryStringConditionConfigMatches(desired *elbv2.QueryStringConditionConfig, current *elbv2.QueryStringConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+func sourceIpConditionConfigMatches(desired *elbv2.SourceIpConditionConfig, current *elbv2.SourceIpConditionConfig) bool {
+	if desired == nil || current == nil {
+		return desired == current
+	}
+	return sliceMatches(desired.Values, current.Values, reflect.DeepEqual)
+}
+
+// stringSliceMatches checks whether current slice matches desired slice, without comparing order.
+func stringSliceMatches(desired []*string, current []*string) bool {
+	if len(desired) != len(current) {
+		return false
+	}
+	visited := make([]bool, len(current))
+	for _, elemI := range desired {
+		found := false
+		for j, elemJ := range current {
+			if visited[j] {
+				continue
+			}
+			if aws.StringValue(elemI) == aws.StringValue(elemJ) {
+				visited[j] = true
+				found = true
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+func sliceMatches(listA interface{}, listB interface{}, comparator func(interface{}, interface{}) bool) bool {
```

Comment:
This function looks fairly slow, how often are the rules checked?

Reply:
the rule check currently triggers on pod relocate events(maybe batched) in v1 version.
This won't be bottleneck since the rules are pretty small. 
In fact, the code logic is much more clean and can be even faster than other comparsion method like sort+compare/hash+compare, given the two list is pretty small and highly likely equal.

Reply:
Makes sense, thanks for the clarification.